clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala

```diff
@@ -200,59 +134,61 @@ class ClioWebClient(
         method = HttpMethods.POST,
         entity = entity
       )
-    ).flatMap(unmarshal[UpsertId])
+    ).via(parserFlow)
   }
 
   def query[CI <: ClioIndex](clioIndex: CI)(
     input: clioIndex.QueryInputType,
     includeDeleted: Boolean
-  ): Future[Json] = {
+  ): Source[Json, NotUsed] = {
     import clioIndex.implicits._
     rawQuery(clioIndex)(input.asJson, includeDeleted)
   }
 
   def getMetadataForKey[CI <: ClioIndex](clioIndex: CI)(
     input: clioIndex.KeyType,
     includeDeleted: Boolean
-  ): Future[Option[clioIndex.MetadataType]] = {
+  ): Source[clioIndex.MetadataType, NotUsed] = {
     import clioIndex.implicits._
     import s_mach.string._
 
     val keyJson = input.asJson
     val keyFields = FieldMapper[clioIndex.KeyType].fields.keySet
       .map(_.toSnakeCase(CirceEquivalentCamelCaseLexer))
 
-    rawQuery(clioIndex)(keyJson, includeDeleted).map { out =>
-      val metadata = for {
-        jsons <- out.as[Seq[Json]]
-        json <- jsons match {
-          case Nil => Right(None)
-          case js :: Nil =>
-            js.mapObject(_.filterKeys(!keyFields.contains(_)))
-              .as[clioIndex.MetadataType]
-              .map(Some(_))
-          case many =>
-            Left(
-              new IllegalStateException(
-                s"""Got > 1 ${clioIndex.name}s from Clio for key:
-                   |${keyJson.pretty(Printer.spaces2)}
-                   |Results:
-                   |${many.asJson.pretty(Printer.spaces2)}""".stripMargin
+    rawQuery(clioIndex)(keyJson, includeDeleted)
+      .fold[Either[Throwable, Option[clioIndex.MetadataType]]](Right(None)) {
+        (acc, json) =>
+          acc match {
+            case Right(None) =>
+              json
+                .mapObject(_.filterKeys(!keyFields.contains(_)))
+                .as[clioIndex.MetadataType]
+                .map(Some(_))
+            case Right(Some(_)) =>
+              Left(
+                new IllegalStateException(
+                  s"""Got > 1 ${clioIndex.name}s from Clio for key:
+                     |${keyJson.spaces2}""".stripMargin
+                )
               )
-            )
-        }
-      } yield {
-        json
+            case _ => acc
+          }
+      }
+      .flatMapConcat {
+        _.fold(
+          Source.failed,
+          // Writing the 'Some' case as just 'Source.single' runs afoul of Scala's
+          // restriction on dependently-typed function values.
+          _.fold(Source.empty[clioIndex.MetadataType])(m => Source.single(m))
+        )
       }
-
-      metadata.fold(throw _, identity)
-    }
   }
 
   private[webclient] def rawQuery(clioIndex: ClioIndex)(
     input: Json,
     includeDeleted: Boolean
-  ): Future[Json] = {
+  ): Source[Json, NotUsed] = {
```

Comment:
Not truly a  raw query since this will get unmarshalled into a QueryInput object right? We will probably add a true raw query endpoint at some point I imagine.